api/rust/core/src/runnable.rs

```diff
@@ -1,90 +1,61 @@
-use crate::STATE;
-use std::mem;
-use std::slice;
+// Since execution is single threaded
+use once_cell::unsync::OnceCell;
 
-extern "C" {
-	fn return_result(result_pointer: *const u8, result_size: i32, ident: i32);
-	fn return_error(code: i32, result_pointer: *const u8, result_size: i32, ident: i32);
-}
-
-pub struct RunErr {
-	pub code: i32,
-	pub message: String,
-}
-
-impl RunErr {
-	pub fn new(code: i32, msg: &str) -> Self {
-		RunErr {
-			code,
-			message: msg.into(),
-		}
-	}
-}
+use crate::env;
+use crate::errors::runnable_unset;
+use crate::errors::RunErr;
 
-pub struct HostErr {
-	pub message: String,
+pub fn return_error(err: RunErr) {
+	let RunErr { code, message } = err;
+	env::return_error(code, message.as_ptr(), message.len() as i32)
 }
 
-impl HostErr {
-	pub fn new(msg: &str) -> Self {
-		HostErr {
-			message: String::from(msg),
-		}
-	}
+pub fn return_result(result_data: Vec<u8>) {
+	env::return_result(result_data.as_ptr(), result_data.len() as i32)
 }
-
 pub trait Runnable {
 	fn run(&self, input: Vec<u8>) -> Result<Vec<u8>, RunErr>;
 }
+pub(crate) fn execute(input: Vec<u8>) {
+	unsafe {
+		STATE
+			.get()
+			.map_or_else(|| Err(runnable_unset()), |state| state.runnable.run(input))
+			.map_or_else(return_error, return_result)
+	}
+}
 
 pub fn use_runnable(runnable: &'static dyn Runnable) {
 	unsafe {
-		STATE.runnable = Some(runnable);
+		if STATE.set(State { runnable, ident: 0 }).is_err() {
+			return_error(RunErr::new(0, &"Can only set runable once"))
+		}
 	}
 }
```

Comment:
What about changing the name to `set_runnable`?

Reply:
Sure! Was going to handle that in a later PR.  But I'll go ahead since this makes it clearer that it's a update and not an action.